### ISSUE TITLE
Change stream quality from within player

### ISSUE THREAD

**Issue by benkomor** [aqw/xbmc-gamepass#137]
Been trying to figure out if it's possible to change the video stream quality through the video player. No luck so far, furthest I get is doing it through the settings button. Any way of doing this during viewing of the game itself - like with the official game pass browser 'app'? If not - would love to do a feature request :-)

Why? Regularly experience latency issues on the sunday evening (located in Amsterdam, NL), probably like all the rest of the folks in Europe. Seems the Game Pass servers are very sensitive to sudden spikes in traffic. It's kind of annoying to exit games / red zone to tweak stream quality each time.

**Comment by BaumSchorle** [comment]
Just to mention it, there is a feature request for the adaptive bitrate stuff. It would be awesome if this would be implemented.

http://forum.kodi.tv/showthread.php?tid=237333

**Comment by emilsvennesson** [comment]
From what I've understood, this is currently only supported for MPEG-DASH (https://github.com/mapfau/inputstream.mpd) and Smooth Streaming (https://github.com/mapfau/inputstream.smoothstream). This plugin uses HLS which there are no input stream add-ons for right now. Perhaps Game Pass is using DASH on other platforms if we try playing around with the User-Agent?

**Comment by Alexqw** [comment]
@emilsvennesson Damn, you're right. I wasn't reading the release notes closely enough.

It's unlikely that Game Pass would support multiple streaming standards, but stranger things have happened.

---Alex

**Comment by kaileu** [comment]
@emilsvennesson no Gamepass using hls on all plattforms at the moment.  
(Xbox, Android, ios, pc)

If we have a good c Programmer we could implement a hls addon.

Mit freundlichen Grüßen
Kai Leutner

**Comment by BaumSchorle** [comment]
I had an idea for a quick and dirty approach as long as HLS is not supported with inputstream and adaptive bitrate.

We could create an xbmc.Playlist object and hand it to the player, this would result in a possibility to quickly switch bitrate from within the player. This would only be possible with live games and Network/Redzone Live, because with archived games it play from beginning when skipping.

I did a short proof of concept of this, worked great as long as one just skips forward. Seems like in Kodi it's not possible to play previous video. :-1:

**Comment by Alexqw** [comment]
@BaumSchorle clever idea. It's a hack, but a nice one. ;-)

IMO, if upstream is going to work on an HLS inputstream, then there's not much sense in going this route, but I don't know what their priorities are. HLS is a /really/ common stream type.

---Alex

**Comment by kaileu** [comment]
The problem is that Kodi devs aren't interessted in hls. And the  
Inputstream Addons are not from the Kodi dev itself.
So i didn't found a project at the moment working on it.

Mit freundlichen Grüßen
Kai Leutner

**Comment by BaumSchorle** [comment]
Yup they created this possibilities to get all this support problems out of Kodi core. The good thing is, that it's possible now, but there has to be a C dev, who want's to work on this.

Maybe we can gather together a group of people who need the HLS inputstream support and can convince a dev to work on this!? I think of the guys who created the MLB and NHL plugin, since they are Neulion as well (at least as far as i know).

**Comment by kaileu** [comment]
Yes @eracknaphobia is also looking for this. But i think he is also a non c Programmer.

But NHL and MLB didn't use neulion anymore they are using the mlb Servers but the addon itself would be the same because the m3u8 Files has the same Body.

**Comment by Alexqw** [comment]
@kaileu Ahh, I just assumed it was done by the core devs as it's included in the default builds.

@BaumSchorle I agree that it would be a good idea to reach out to the other addons which use HLS. They already asked (somewhat) https://github.com/mapfau/inputstream.mpd/issues/13

---Alex

**Comment by grooves12** [comment]
Is there any way possible to pull in the 60fps high-bitrate streams that are currently only available on the Xbox?

**Comment by aqw** [comment]
Anthony Waters has posted a first-stab at implementing an HLS input stream for Kodi. :-D

https://github.com/awaters1/inputstream.hls

According to him, he has only tested it on his own Linux desktop (hasn't tested Windows, android, anything else) with his ESPN3 addon. But it does work!

I don't have time to play with this right now, but I figured I'd post it here in case someone feels motivated this long weekend. :-D

---Alex

**Comment by eracknaphobia** [comment]
That's great! Thanks for sharing.

**Comment by emilsvennesson** [comment]
@peak3d has started working on HLS support in inputstream.adaptive now as well, so that's definitely something to keep an eye out for. :-)

**Comment by aqw** [comment]
@emilsvennesson Fantastic. Thanks for the info. :-)

---Alex

**Comment by peak3d** [comment]
thanx @emilsvennesson for letting me know, I'm currently at the beginning of HLS but can stopp my work if the addon from @awaters1 is doing a good job!

1.) Is it intended to push the addon to kodi mainline?
2.) Does this addon work with the official kodi wnbc addon's m3u format?

**Comment by awaters1** [comment]
@peak3d I have only tried it with the ESPN3 addon that uses mpegts within the HLS stream.  It was my first binary addon so I didn't have any plans initially to have it mainlined.  With that in mind the one in the master branch works in my limited testing but lacks smart adaptive stream switching.  I have a new branch stream-switching which adds in much improved switching based on reinforcement learning, in my testing it works pretty well.  There is however, one caveat, the stream switching doesn't work reliably.  Sometimes when switching to different renditions Kodi will freeze the video while the audio keeps playing and then do a seek backwards in time to correct.  I have tried different techniques to push the `DMX_SPECIALID_STREAMCHANGE` but from my limited understanding of Kodi I wasn't able to figure something out to prevent the freezes.  It seems as though as soon as a `DMX_SPECIALID_STREAMCHANGE` packet is read by `DemuxRead` then Kodi calls `GetStreamIds`, which will return the new stream information but there are packets still in the demux cache referring to the old stream information so the new demuxer uses the new stream information with the old packets, then when the new packets hit the video resolution is changed which appears to mess things up sometimes. I tried to look through Kodi's code to figure out if I am doing things correctly but came up empty handed, except for some references to EnableStreamAtPTS and having multiple video demuxers, neither of which I was able to get working.

I'll look into WNBC to see if it works, the HLS parsing and mpegts demuxing are simple, it is the segment downloading, playlist reloading and stream switching that caused things to get complex.
